FABGW-21: Realtime implementation of ChaincodeEvents service

bestbeforetoday opened the pull request:
Initial implementation of Fabric Gateway's ChaincodeEvents service, allowing only listening for realtime chaincode events.

Also add block number into CommitStatus response.

bestbeforetoday committed:
FABGW-21: Realtime implementation of ChaincodeEvents service

Initial implementation of Fabric Gateway's ChaincodeEvents service, allowing only listening for realtime chaincode events.

Also add block number into CommitStatus response.

Signed-off-by: Mark S. Lewis <mark_lewis@uk.ibm.com>

C0rWin commented on internal/pkg/gateway/commit/commitadapter.go:
```diff
@@ -28,13 +27,32 @@ func (adapter *PeerAdapter) CommitNotifications(done <-chan struct{}, channelNam
+	}
+	if block.Header == nil {
+		return nil, errors.Errorf("missing header for block containing transaction %s", transactionID)
+	}
```
This appears a bit weird to me, as you check whenever the header is not nil, while not checking whenever a block is not nil. not something expected to happen here. However, since you are already adding such a check for the header, I would expect to have a similar check for the block itself.

bestbeforetoday replied:
Won't `GetBlockByTxID()` return a non-nil error if there is no block?

C0rWin replied:
Well, it will also return a valid non-nil header in case there are no errors. I guess it depends on what was your intent while adding this check for the non-nil header.

See: 

```go
func deserializeBlock(serializedBlockBytes []byte) (*common.Block, error) {
	block := &common.Block{}
	var err error
	b := newBuffer(serializedBlockBytes)
	if block.Header, err = extractHeader(b); err != nil {
		return nil, err
	}
	if block.Data, _, err = extractData(b); err != nil {
		return nil, err
	}
	if block.Metadata, err = extractMetadata(b); err != nil {
		return nil, err
	}
	return block, nil
}
```

How the block is actually being fetched. Now, this is a nit comment, not a real show stopper, just for the sake of consistency and clarity. Like, why did you check for non-nil header?

bestbeforetoday replied:
It's there because any time I deliver code that deals with protobufs and doesn't explicitly check for nil, it gets flagged in review comments. I'm quite happy to remove the check if that's the preferred approach.

bestbeforetoday replied:
I've pushed an update that removes the nil check on the block header

C0rWin replied:
My initial comment and intent were not to remove the check, but for sanity and clearness, add the check to ensure the block is not nil neither. ;)

bestbeforetoday replied:
I'm not convinced we can reasonably expect `GetBlockByTxID()` to return `nil, nil`. The error response is there to indicate that a block isn't returned. It seems confusing and non-idiomatic to check both the value and the associated error response for nil.

bestbeforetoday commented:
/ci-run

bestbeforetoday commented:
Refactored to break cyclic dependency and enable the use of Counterfeiter as the mocking framework again.

manish-sethi commented on internal/pkg/gateway/commit/finder.go:
```diff
@@ -30,26 +36,34 @@ type Finder struct {
 	}
 
-	if status, err := finder.Query.TransactionStatus(channelName, transactionID); err == nil {
+	if code, blockNumber, err := finder.Query.TransactionStatus(channelName, transactionID); err == nil {
```
Though not related to this PR but just noticed - In the case of `err != nil`, it would proceed and wait on channel, even if the transaction is already committed (and in this scenario, eventually, the client would receive irrelevant error). I would assume that the preferred behaviour should be to return on  error.

bestbeforetoday replied:
There should only be an error here if the transaction does not exist (i.e. is not already committed), in which case we want to wait for it to be committed, not return an error

manish-sethi replied:
That's not correct. This function internally queries the ledger and there are four situations that can cause this error.
- TxID not present (as you mentioned) 
- TxID is present but the related validation code is not present (this could happen if the peer is bootstrapped from a snapshot)
- Any I/O error while accessing ledger

bestbeforetoday replied:
OK, I buy that `GetTxValidationCodeByTxID` may return an error for situation in which it makes more sense to return an error. How can I distinguish between them? What I definitely can't do is never wait for transactions that have not yet been committed, so I either need a completely robust way of distinguishing the different error conditions, or just go ahead with the listening anyway and hope for the best. The worst that happens is eventually the client gives up waiting and I stop listening anyway.

manish-sethi commented on internal/pkg/gateway/commit/commitadapter.go:
```diff
@@ -28,13 +28,28 @@ func (adapter *PeerAdapter) CommitNotifications(done <-chan struct{}, channelNam
+		return 0, 0, err
+	}
+
+	block, err := ledger.GetBlockByTxID(transactionID)
```
As far as I recall, you wanted to include the block number here for the logging purpose. However, this introduces a significant overhead just to get the block number. 

If you really want to include the block number, my suggestion would be that we enhance the API `GetTxValidationCodeByTxID` such that it includes the block number as well -- In the [index](https://github.com/hyperledger/fabric/blob/main/common/ledger/blkstorage/blockindex.go#L130) for `txID`, we include the block number as well.

bestbeforetoday replied:
Logging and debugging are the primary uses. It might also be useful as a constraint for future interactions, since we know the minimum block height required to see this transaction, but the Gateway is not making use of this today so not a showstopper.

Being able to grab just the block number out of the ledger index would be great, either along with the transaction ID or separately. If that's not practical then I'll give up on including the block number along with the validation code so I can make some progress on chaincode eventing, which also requires changes to this code.

manish-sethi replied:
Yes, I was thinking on the lines of enhancing `GetTxValidationCodeByTxID` to return the block number along with the validation code. However, your PR will have to wait for that change. So one option could be to do that outside this PR. I leave the choice to you. If you prefer to do in this PR, I can spend some time on enhancing the above API

bestbeforetoday replied:
Now using the new implementation to get block number directly from the `GetTxValidationCodeByTxID` call

manish-sethi commented on integration/gateway/gateway_test.go:
```diff
@@ -238,13 +238,11 @@ var _ = Describe("GatewayService", func() {
-			}
-			Expect(proto.Equal(actualStatus, expectedStatus)).To(BeTrue(), "Expected\n\t%#v\nto proto.Equal\n\t%#v", actualStatus, expectedStatus)
+			Expect(status.Result).To(Equal(peer.TxValidationCode_VALID))
+			Expect(status.BlockNumber).To(BeNumerically(">", 0))
```
Better to use the exact block number here.

bestbeforetoday replied:
How do we know what the block number will be without introducing a fragile dependency on other integration tests and their ordering?

manish-sethi replied:
Not sure I understand the scope when you say other integration tests but before `Describe("CommitStatus"`, it sets up a fresh network...no?

bestbeforetoday replied:
My worry is, what is the block number I should pick? And are we guaranteed that will not change in the future, either because of changes to Fabric implementation or because somebody comes along and adds another test case to this Describe block? I think I'd prefer to test that some sane block number came back rather than deliver fragile tests that may break in the future for reasons not connected to this particular test. It's a moot point for now as we're not happy with the approach for obtaining the block number so I'll remove the assertion for now

bestbeforetoday replied:
I've changed the test to check the commit status of two subsequent transactions so I can test that the block number of the second commit is one higher than the block number of the first

bestbeforetoday commented:
@manish-sethi Moving back to draft to continue changes to support chaincode event listening, at least until there is a suitable mechanism to grab block number along with transaction status and a satisfactory interim PR can be delivered.

manish-sethi commented:
> @manish-sethi Moving back to draft to continue changes to support chaincode event listening, at least until there is a suitable mechanism to grab block number along with transaction status and a satisfactory interim PR can be delivered.

@bestbeforetoday - The [PR](https://github.com/hyperledger/fabric/pull/2614) for block number is merged now.

manish-sethi commented on internal/pkg/gateway/api.go:
```diff
@@ -258,11 +258,55 @@ func (gs *Server) CommitStatus(ctx context.Context, signedRequest *gp.SignedComm
+		return status.Error(codes.InvalidArgument, "a chaincode events request is required")
+	}
+
+	request := &gp.ChaincodeEventsRequest{}
+	if err := proto.Unmarshal(signedRequest.Request, request); err != nil {
+		return status.Error(codes.InvalidArgument, "invalid chaincode events request")
+	}
```
Even if there is no error, this could still leave `request` with one or more empty field (e.g., `signedRequest.Request` could be an empty bytes). Better to explicitly check for this so it does not lead to nil-pointer panic in the downstream code. The downstream code may already be checking for this but it becomes hard to reason about at higher level so better to have independent check at the entry level.

I guess that this applies to existing function in this file as well.

bestbeforetoday replied:
How will an empty string or empty slice cause a nil-pointer panic? Could you give me an example of the change you are requiring here?

manish-sethi replied:
In downstream code this byte array is used to unmarshal and I am not certain that every place all the checks have been in place. So, it was just to fail early if this byte slice is nil - kind of basic input check.

bestbeforetoday replied:
You can unmarshall a zero-length `[]byte` fine. I don't understand the benefit of adding code bloat for a special check for a zero-length slice rather than letting a meaningful error be raised, such as the requested channel doesn't exist or the supplied identity is not valid. The client APIs don't send empty protobufs so it's not a scenario that can occur by accident

manish-sethi commented on internal/pkg/gateway/commit/blocknotifier.go:
```diff
@@ -0,0 +1,76 @@
+	done          <-chan struct{}
+	lock          sync.Mutex
+	listeners     []blockListener
+	closed        bool
```
nit: convention - lock just above what it protects (in other file as well)
```suggestion
	listeners     []blockListener
	
	lock          sync.Mutex
	closed        bool
```

manish-sethi commented on internal/pkg/gateway/commit/chaincodeeventnotifier.go:
```diff
@@ -0,0 +1,166 @@
+	Events      []*peer.ChaincodeEvent
+}
+
+type chaincodeEventListenerSet map[*chaincodeEventListener]bool
```
nit: standard usage for `set`  (in other file as well)
```suggestion
type chaincodeEventListenerSet map[*chaincodeEventListener]struct{}
```

manish-sethi commented on internal/peer/node/start.go:
```diff
@@ -825,14 +825,18 @@ func serve(args []string) error {
+					},
+					&commit.Eventer{
+						Notifier: notifier,
 					},
```
Though not introduced in this PR but it got further flared - This pattern of linking the dependencies surfaces much of internal details of gateway module to this level. The only real dependency of gateway module is a few functions on the `peerInstance`. I feel that this will be clearer if you could define a limited interface for this and just pass `peerInstance` in `gateway.CreateServer` and construct `PeerAdapter`, `notifier`, `Finder`, and `Eventer` internally inside gateway module. This may also remove many interfaces in the gateway module.

bestbeforetoday replied:
I started with the narrower interface but was required to do it this way by another maintainer. The way the code is currently structured (at the request of another maintainer), we can't pass concrete implementations in to `gateway.CreateServer` since mocks are passed in there by unit tests.

I'll look at some changes back towards my original approach for the `Finder` and `Eventer`, and I propose we look at changes to other areas of the code in a different PR rather than shoe-horn a bunch of unrelated code refactor into this one.

bestbeforetoday replied:
I've made some changes to reduce the leakage of internal abstractions outside of the gateway package, which I think is along the lines of what you are asking.

Note that even within the gateway packages there are deliberately different layers of abstraction. The core Gateway code is intended to be agnostic to whether it is an embedded or standalone implementation, and so interacts with the embedded implementation-specific elements using interfaces, which could be replaced standalone implementations.

manish-sethi replied:
We can definitely do it in a separate PR. But I feel that there are way to many layers of abstraction for a simple enough functionality.

manish-sethi replied:
> we can't pass concrete implementations in to gateway.CreateServer

contrary to this you were actually passing the peer instance to `commit package`. What my suggestion was that only external dependency of gateway package is to get an object on which it can call "RegisterForNotification" for a given channel and that should be easy enough to reflect in first-order via an interface.

bestbeforetoday replied:
> What my suggestion was that only external dependency of gateway package is to get an object on which it can call "RegisterForNotification" for a given channel and that should be easy enough to reflect in first-order via an interface.

No.

`RegisterForNotification()` implies the peer embedded implementation, which is in the commit package. The commit package uses the `NotificationSupplier` interface to define `CommitNotifications()` which is the same as `RegisterForNotification()`, so it's already doing as you suggest.

The gateway server is not intended to be tied to either a peer embedded or standalone implementation, so it uses a `CommitFinder` interface to obtain transaction validation codes by some means that it doesn't need to know about, and which would be significantly different in the embedded and standalone implementations.

manish-sethi commented on internal/pkg/gateway/commit/eventer.go:
```diff
@@ -0,0 +1,17 @@
+
+func (e *Eventer) ChaincodeEvents(ctx context.Context, channelName string, chaincodeName string) (<-chan *BlockChaincodeEvents, error) {
+	return e.Notifier.notifyChaincodeEvents(ctx.Done(), channelName, chaincodeName)
+}
```
Just to confirm my understanding, you plan to add more function into this struct for delivering events from historical blocks. Otherwise, in the current form, it looks an indirection that could be avoided.

bestbeforetoday replied:
Correct

manish-sethi commented on internal/pkg/gateway/commit/finder.go:
```diff
@@ -12,13 +12,20 @@ import (
 // QueryProvider provides status of previously committed transactions on a given channel. An error is returned if the
 // transaction is not present in the ledger.
 type QueryProvider interface {
-	TransactionStatus(channelName string, transactionID string) (peer.TxValidationCode, error)
+	TransactionStatus(channelName string, transactionID string) (peer.TxValidationCode, uint64, error)
 }
```
To highlight the comment that I made in `start.go` in context - This is how it's easy for someone to get lost in the maze of layers of finer-grained dependencies as opposed having a simplified straight-forward dependency. In this case, this interface, it's sole implementation, and it's sole consumer all three are in the same package. Which makes it unnecessarily round-about to understand the true linkages.

bestbeforetoday replied:
This describes an interface for a subset of capability provided by the PeerAdapter struct, which allows me to use a mock implementation in unit tests. Trying to make the PeerAdapter work purely with interfaces (that I could then mock) would be even more convoluted because of the way to Go type system works

manish-sethi replied:
Decoupling code within a package for the sake of using mocks for UTs does not sound a good idea to me. I am not sure I understand what prevents from writing UTs as long as external dependencies are mocked.

bestbeforetoday replied:
QueryProvider decouples the PeerAdapter (i.e. the thin layer of code not easily untangled from the peer implementation, and so you could think of as being the external dependency) from the implementation in this package

manish-sethi commented on internal/pkg/gateway/commit/blocknotifier.go:
```diff
@@ -0,0 +1,76 @@
+type blockListener interface {
+	receiveBlock(*ledger.CommitNotification)
+	close()
+}
```
Can we remove this interface and straight away use the two concrete types - to make it more explicit. The current code may have saved a few lines of code but lessen the readability.

bestbeforetoday replied:
This abstraction isn't there to save a few lines of code. It's to:
1. Keep the blockListener [focused on the task](https://en.wikipedia.org/wiki/Single-responsibility_principle) of receiving and dispatching events from the ledger
2. Allow it to be [extended with other listeners without requiring code changes](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle)
3. Avoid it needing [dependencies on listener implementations](https://en.wikipedia.org/wiki/Dependency_inversion_principle)

manish-sethi replied:
These objectives sounds good if they are between course level modules. Applying these to micro level of granularity has led to Fabric code at many places as hard to read IMO.

bestbeforetoday replied:
Feel free to make that change

manish-sethi commented on internal/pkg/gateway/commit/blocknotifier.go:
```diff
@@ -0,0 +1,76 @@
+	close()
+}
+
+type blockNotifier struct {
+	commitChannel <-chan *ledger.CommitNotification
```
I feel that too many `notifiers` are making it hard to read code. Can we club the functionality in this into the main `Notifier`. That will make the main `Notifier` act as a central piece that glues together the lower level events received from ledger and the other two notifiers handle the event specific delivery.

bestbeforetoday replied:
There is a one-to-many relationship between `Notifier` and `blockNotifier`. Also `Notifier` needs to be able to throw away and recreate a `blockNotifier`. I'm not sure how you see merging `Notifier` and `blockNotifier` working

manish-sethi replied:
I was thinking in the lines of launching a gorouting for each channel directly into the `Notifier` and the notifier just keeps track of the two event specific notifiers. In the current structure of the code it sounded like all three notifiers are kind of siblings when they are actually not. The other approach could be to have the both event specific notifiers only inside this `blocknotifier` - so this becomes full controller of channel specific notifier and hierarchy is clearer.

bestbeforetoday replied:
Feel free to make that change

manish-sethi reviewed (commented):
@bestbeforetoday - I feel that the code in this area is already becoming complex enough. Primarily because of how dependencies are factored out and a number of notifiers. I have listed a few suggestions to simplify things, see if they makes sense to you.

bestbeforetoday commented:
/ci-run

bestbeforetoday committed:
Refactor to minimise visibility of implementation outside package

Previously node.serve() accessed a lot of implementation detail within the gateway package to create structs of the appropriate abstraction for the gateway internals. Refactored to move the translation between external structs and internal abstractions to the gateway.CreateServer() function, and moved the old implementation to an newServer() function that is used by unit tests to create a Gateway server that is agnostic to whether it is an embedded or standalone implementation.

Signed-off-by: Mark S. Lewis <mark_lewis@uk.ibm.com>

bestbeforetoday commented:
@manish-sethi Your outstanding comments seem to be primarily objections to code structure that was already there prior to this PR, and was there at the request of other maintainers. I have addressed as many of these structural issues as I think I reasonably can in this PR, which is delivering chaincode event listening, not a refactor of existing code

manish-sethi commented on integration/gateway/gateway_test.go:
```diff
@@ -129,175 +187,104 @@ var _ = Describe("GatewayService", func() {
-			badSigningIdentity := network.OrdererUserSigner(network.Orderer("orderer"), "Admin")
-			badIdentity, err := badSigningIdentity.Serialize()
+	Describe("ChaincodeEvents", func() {
+		It("should respond with emitted chaincode events", func() {
```
I guess that you still plan to add security related tests, like the one in `CommitStatus` api ? However, that can always be added later.

manish-sethi commented:
> @manish-sethi Your outstanding comments seem to be primarily objections to code structure that was already there prior to this PR, and was there at the request of other maintainers. I have addressed as many of these structural issues as I think I reasonably can in this PR, which is delivering chaincode event listening, not a refactor of existing code

Yes, As I mentioned in the previous comment, the refactoring can be taken up in a separate PR

bestbeforetoday commented:
@manish You were not keen on having the `blockNotifier` in addition to the other notifiers. This is really there because you imposed the limitation that there could only be one listener to ledger notifications, so the `blockNotifier` provides a mechanism to workaround that limitation and allow multiple listeners. One approach to simplifying this code might be to remove your restriction on only allowing a single listener for ledger notifications, in which case the `blockNotifier` and the logic there could be completely removed.
